### PR TITLE
Remote Signing

### LINKED ISSUE #7: Protected header should not be re-serialized for verification.
The verification process of COSE Sign1 signature is as below.

```go
msg := &cose.Sign1Message{}
err := cbor.Unmarshal(sign, msg)
check(err)

err = msg.Verify(nil, *verifier)
check(err)
```

To verify the message, we need to unmarhsal the binary data to a `Sign1Message` structure. In the unmarhsal process, the protected header (encoded as a binary string) is deserialized into a map.

Later in `msg.Verify()` calls `EncodeProtected()` to re-serialized the protected header map back to a binary string for computing the digest-to-be-verified. Although canonical encoding is applied, the re-serialized protected header may not be the same as the original header, especially the header is encoded with tags. Hence, the verification will fail.

We should refactor the `go-cose` library so that the original protected header in the binary string form is used in the digest computation withour re-serialization.

### LINKED ISSUE #8: `nil` external data should imply empty slice, not nil
In [RFC8152 4.4](https://datatracker.ietf.org/doc/html/rfc8152#section-4.4), the step 4 states

> The protected attributes from the application encoded in a bstr type.  If this field is not supplied, it defaults to a zero-length binary string.

```
Sig_structure = [
    context : "Signature1",
    body_protected : empty_or_serialized_map,
    external_aad : bstr,
    payload : bstr
]
```

That is, when `external` is `nil` for `Sign` and `Verify`, it should default to `[]byte{}` so that a proper `Sig_Structure` can be constructed. Otherwise, the `go-cose` library will fail all verifications agaist signatures generated by other libraries.

Note: `[]byte{}` is `0x40` and `nil` is `0xf6` in CBOR.

### LINKED ISSUE #9: Key extensibility support
Currently, the siging key and the verification key must be a golang built-in type (e.g. `rsa.PrivateKey`).

We need to refactor the `go-cose` library so that `Signer` and `Verifier` are interfaces instead of solid structures. With those interface, other developers can make remote signing keys possible and even support FIPS 140-2 by using OpenSSL via CGO.

### LINKED ISSUE #10: More algorithms support
Although there are lots of algorithms are defined in [algorithms.go](https://github.com/veraison/go-cose/blob/master/algorithms.go), most of them are not implemented.

Since COSE is often compared with JOSE, we would like support the following additional algorithms in `go-cose`, with test coverage:

- `EdDSA`
- `PS384`
- `PS512`

[Notary algorithms](https://github.com/notaryproject/notaryproject/blob/main/signature-specification.md#signature-algorithm-requirements)

### PR DESCRIPTION
**Pull request by shizhMSFT**
Objectives:
- Resolves #7
- Resolves #8
- Resolves #9
- Resolves #10

### CODE REVIEW

**Comment by qmuntal** [comment]
This is another attempt to simplify the `Algorithm` data model by using constant identifiers, yet allowing custom Algorithms implementations:

```go
type Algorithm int

type extAlgorithm struct {
	Name     string
	HashFunc crypto.Hash
}

var extAlgorithms = make(map[Algorithm]extAlgorithm)

const (
	PS256 Algorithm = -37
	ES256 Algorithm = -7
	ES384 Algorithm = -35
	ES512 Algorithm = -36
)

func RegisterHash(id Algorithm, name string, h crypto.Hash) {
	extAlgorithms[id] = extAlgorithm{name, h}
}

func (a Algorithm) Name() string {
	switch a {
	case PS256:
		return "PS256"
	case ES256:
		return "ES256"
	case ES384:
		return "ES384"
	case ES512:
		return "ES512"
	default:
		if x, ok := extAlgorithms[a]; ok {
			return x.Name
		}
		return "unknown algorithm value " + strconv.Itoa(int(a))
	}
}

func (a Algorithm) String() string {
	return a.Name()
}

func (a Algorithm) HashFunc() crypto.Hash {
	switch a {
	case PS256:
		return crypto.SHA256
	case ES256:
		return crypto.SHA256
	case ES384:
		return crypto.SHA384
	case ES512:
		return crypto.SHA512
	default:
		if x, ok := extAlgorithms[a]; ok {
			return x.HashFunc
		}
		panic("go-cose: unknown algorithm value " + strconv.Itoa(int(a)))
	}
}
```

Discussed offline with @shizhMSFT

File: headers.go
Comment (qmuntal) [inline headers.go:66]: Go tends to avoid prefixing getters with `Get`: https://go.dev/doc/effective_go#Getters. `Algorithm()` feels more idiomatic,

File: headers.go
Comment (qmuntal) [inline headers.go:69]: [RFC 8152 Section 16.4](https://www.rfc-editor.org/rfc/rfc8152.html#section-16.4) says the value 0 is reserved, can't be used. With this in mind, does it makes sense to not return an error here, just `return 0, nil`? I'm saying this because it is not mandatory for protected headers to contain an algorithm parameter, so it seem weird to return an error.

Comment (shizhMSFT) [reply headers.go:69]: The value `0` is valid value even if it is reserved. Therefore, we should return an error for that.

```go
alg, err := header.GetAlgorithm()
if err == nil {
   // do stuff with the alg
} else if err != ErrAlgorithmNotFound {
   panic(err)
} else {
  // alg does not present.
}
```

With above code, the caller can know whether the `alg` header present or not not.

If we `return 0, nil`, then "`alg` not present" and "`alg` is `0 `" are not distinguishable.

File: algorithm.go
Comment (shizhMSFT) [reply algorithm.go:113]: Actually, I am thinking about dropping
```golang
func (a Algorithm) HashFunc() crypto.Hash
```
as it is not used anywhere.

File: headers.go
Comment (qmuntal) [reply headers.go:69]: > The value 0 is valid value even if it is reserved. Therefore, we should return an error for that.

That's a pitty. Then the error sentinel is the way to go! 👍

File: ecdsa.go
Comment (qmuntal) [inline ecdsa.go:19]: Nice API improvement!

File: signer.go
Comment (qmuntal) [inline signer.go:65]: Passing `rand io.Reader` from outside rather than using the global `rand.Reader` would ease getting deterministic keys.

File: signer.go
Comment (qmuntal) [inline signer.go:63]: I would be more specific here. Can this function be used in production code?

File: signer.go
Comment (shizhMSFT) [reply signer.go:65]: No magic "deterministic keys" by `randutil.MaybeReadByte()`. Details see https://github.com/golang/go/issues/38548.

File: signer.go
Comment (shizhMSFT) [reply signer.go:63]: If this function is not suggested in production code, then we should remove it. I guess we need more input from others.

File: signer.go
Comment (qmuntal) [reply signer.go:65]: Not for production code, but for testing purposes it is common to use a "zero source", where all the generated random numbers are just 0, and this case is not affected by `randutil.MaybeReadByte()` non-determinism.

File: signer.go
Comment (qmuntal) [reply signer.go:63]: Agree!

File: algorithm.go
Comment (qmuntal) [inline algorithm.go:12]: Types are only propagated when using `iota`. In this const block only `AlgorithmPS256` is an Algorithm, others are untyped constants.

**Comment by qmuntal** [comment]
I'm wondering if we should treat all integers as `int64` instead of the native `int` to avoid overflows in 32 bits platforms. In fact `fxamacker/cbor` already decodes integers as `int64`, but we are then casting them to `int`s.

Among others, this would mean:
- Changing `type Algorithm int` to `type Algorithm int64`
- Changing all `HeaderLabel*` from untyped constant to `int64`

File: signer.go
Comment (thomas-fossati) [reply signer.go:85]: I'm OK for dropping it.

File: ecdsa.go
Comment (qmuntal) [inline ecdsa.go:12]: `golang.org/x/crypto/cryptobyte` does not appear in the `go.mod` file. Run `go mod tidy` to fix it.

File: signer.go
Comment (shizhMSFT) [reply signer.go:43]: > It is possible (but highly unlikely) that someone implements a crypto.Signer whose Public method returns a `rsa.PublicKey`.

Then it is a programming error. The public key type of a RSA key must be `*rsa.PublicKey`.
See the similar implementation in [crypto/x509](https://github.com/golang/go/blob/go1.18/src/crypto/x509/x509.go#L1364-L1393).

Comment (shizhMSFT) [reply signer.go:43]: > Could we case `key` directly to a `rsa.PrivateKey`, which also provides `N`?

`rsa.PrivateKey` is a `rsa.PublicKey` (see https://github.com/golang/go/blob/go1.18/src/crypto/rsa/rsa.go#L101).

Its [Public()](https://github.com/golang/go/blob/go1.18/src/crypto/rsa/rsa.go#L110-L113) method just returns the public key.

File: ecdsa.go
Comment (shizhMSFT) [reply ecdsa.go:12]: I will fix all the `go.mod` when I finish all the coding. I think I just have `Verifier` and `COSE_Sign` left.

**Comment by shizhMSFT** [comment]
> I'm wondering if we should treat all integers as `int64` instead of the native `int` to avoid overflows in 32 bits platforms. In fact `fxamacker/cbor` already decodes integers as `int64`, but we are then casting them to `int`s.
> 
> Among others, this would mean:
> 
> * Changing `type Algorithm int` to `type Algorithm int64`
> * Changing all `HeaderLabel*` from untyped constant to `int64`

Fixed with c231d48.

File: signer.go
Comment (shizhMSFT) [reply signer.go]: `NewSignerWithEphemeralKey` dropped by a163210.

File: signer.go
Comment (shizhMSFT) [reply signer.go]: `NewSignerWithEphemeralKey` was dropped.

File: headers.go
Comment (qmuntal) [inline headers.go:71]: `value` won't be a `int64`, but a `cose.Algorithm`, if the algorithm has been set like this:

```go
headers := cose.Headers{
		Protected: cose.ProtectedHeader{
			cose.HeaderLabelAlgorithm: cose.AlgorithmES512,
		},
	}
```

I've just realized that with the current `Algorithm` definition it will be hard to not mix `int64` and `Algorithm` here and there. Another example is `ProtectedHeader.SetAlgorithm`, which castes the algorithm to a `int64`. And probably when unmarshalling a header we will also have an `int64`.

To solve this problem, what if we change the `Algorithm` definition to a be an alias instead of a type, so we always deal with `int64` types:

```go
type Algorithm = int64
```

The drawback is that we lose the `String` method, but it was not really being used anyway.

Comment (shizhMSFT) [reply headers.go]: We cannot do `type Algorithm = int64`. Otherwise, we are not able to define any methods for this type.

Instead, I store `Algorithm` directly in the map (see 06233d7), and it will be marshaled to int64 anyway.

Comment (qmuntal) [reply headers.go]: Great. Then I would also change `ProtectedHeader.UnmarshalCBOR` so it also cast the algorithm entry (key=1) into an `Algorithm` if it is present.

File: example_sign1_test.go
Comment (qmuntal) [inline example_sign1_test.go:1]: For what I've seen in the wild, people tend to put all examples in the same file, normally named `examples_test.go`, so examples don't pollute the file tree.

File: sign1.go
Comment (qmuntal) [inline sign1.go:121]: @shizhMSFT have you thought about [this proposal](https://github.com/veraison/go-cose/pull/14#discussion_r836487473)?

IMO promoting the Sign and Verify method to a top-level function would make the API easier to use and have less footguns.

```go
func Sign1(rand io.Reader, signer Signer, payload, external []byte, headers Headers) (*Sign1Message, error)
func Verify1(msg *Sign1Message, verifier Verifier) error 
```

File: example_sign1_test.go
Comment (shizhMSFT) [reply example_sign1_test.go:1]: It depends. One counterexample is https://github.com/golang/go/tree/master/src/sort

File: sign1.go
Comment (shizhMSFT) [reply sign1.go:121]: I'm still thinking about what APIs look like for `COSE_Sign` since it should be consistent with `COSE_Sign1`.

Comment (shizhMSFT) [reply sign1.go:121]: `COSE_Sign` has complex signing and verification scenarios. I will drop `Sign` and `Verify` but will adopt `Sign1` and `Verify1`.

File: example_sign1_test.go
Comment (shizhMSFT) [reply example_sign1_test.go]: merged all examples to `example_test.go`.

File: DESIGN.md
Comment (qmuntal) [inline DESIGN.md:174]: Is this function intended to be called by users, or just internally? If the later, we can leave it unexported.

File: sign.go
Comment (qmuntal) [inline sign.go:32]: COSE_Signature does not have an external data property. I also recall having this conversation and saying that there should only be on external data per signed message, regardless of the signature count.

**Comment by SteveLasker** [comment]
@qmuntal, @yogeshbdeshpande, looks like @shizhMSFT is code complete, and ready for more feedback.  What would it take to run the tests, to see how the code and the tests hold up as we merge into the remote signing branch?

Let's also see if we can iterate quickly to merge into the remote-signing branch, then open issues. This will enable notary to start testing as a pointer to this branch.

**Comment by yogeshbdeshpande** [comment]
@SteveLasker I will review the changes tomorrow and will attempt to provide my feedback by end of day 05th April.

File: DESIGN.md
Comment (yogeshbdeshpande) [inline DESIGN.md:184]: Could API on line 184 be changed to be more clear to say:
'GetAlgorithm()'

File: DESIGN.md
Comment (yogeshbdeshpande) [inline DESIGN.md:177]: Could you please add a good comment on the function with acronyms especially:
I2OSP and OS2IP to clarify the purpose of these API's ?

**Comment by qmuntal** [comment]
Other than two minor comments, this looks great!

File: DESIGN.md
Comment (shizhMSFT) [reply DESIGN.md:184]: @qmuntal proposed `Algorithm()` with a reasonable motivation: https://github.com/veraison/go-cose/pull/17#discussion_r837799695.

File: ecdsa_test.go
Comment (shizhMSFT) [reply ecdsa_test.go]: cleaned up

File: DESIGN.md
Comment (shizhMSFT) [reply DESIGN.md:176]: added more comments for the purpose for those APIs.

File: DESIGN.md
Comment (shizhMSFT) [reply DESIGN.md]: I've made it hidden :D

File: cbor.go
Comment (qmuntal) [inline cbor.go:9]: Are these constants intended for user consumption? If not, we can hide them :)

File: headers.go
Comment (qmuntal) [inline headers.go:47]: This function should check if the `alg` header is present and, if so, cast its value to an `Algorithm`.

File: sign1.go
Comment (qmuntal) [inline sign1.go:215]: nit: std library places the payload before the external data in function signatures. Example: https://pkg.go.dev/crypto/cipher#AEAD. We could follow this convention to avoid mistakes for users using go-cose and std.

File: ecdsa.go
Comment (qmuntal) [inline ecdsa.go:155]:
```suggestion
	if !ecdsa.Verify(ev.key, digest, r, s) {
		return ErrVerification
	}
```

File: rsa.go
Comment (qmuntal) [inline rsa.go:55]:
```suggestion
	if rsa.VerifyPSS(rv.key, hash, digest, signature, &rsa.PSSOptions{
		SaltLength: rsa.PSSSaltLengthEqualsHash, // defined in RFC 8230 sec 2
	}) != nil {
```

File: ed25519.go
Comment (qmuntal) [inline ed25519.go:44]:
```suggestion
	if !ed25519.Verify(ev.key, digest, signature) {
```

File: ecdsa.go
Comment (qmuntal) [inline ecdsa.go:97]: We depend on the `x/crypto` module just for reading these ASN.1 encoded big integers. I don't have a strong opinion here, but I would probably just use the build-in `encoding/asn1` package and remove the `x/crypto` dependency:

```go
func decodeECDSASignatureASN1(sig []byte) (r, s *big.Int, err error) {
    var esig  struct {
        R, S *big.Int
    }
    if _, err := asn1.Unmarshal(sig, &esig); err != nil {
        return nil, nil, errors.New("ecdsa: invalid signature: invalid ASN.1 encoding")
    }
    return esig.R, esig.S, nil
}
```

Go crypto package does use it, but I believe the main driver is to avoid making `crypto` depend on `encoding\asn1` for the sake of having a sane dependency tree.

File: sign.go
Comment (yogeshbdeshpande) [inline sign.go:155]: .Nit: Typo on line 155: contructs

File: sign.go
Comment (yogeshbdeshpande) [reply sign.go:32]: I agree with @qmuntal here!

File: ecdsa.go
Comment (qmuntal) [inline ecdsa.go:25]: RFC8017 Section 4.1 cases says:
> If x >= 256^xLen, output "integer too large" and stop.

`if len(x.Bits()) > len(buf) {` does not fully implement this check.

For example, the following invalid inputs are not catch by the if-check and result in a runtime panic inside `FillBytes`:

```go
cose.I2OSP(big.NewInt(256*256*256), make([]byte, 3))
```

I think this should do the trick:
```suggestion
	if (x.BitLen()+7)/8 > len(buf) {
```

**Review by qmuntal — CHANGES_REQUESTED** [review]

File: rsa.go
Comment (shizhMSFT) [reply rsa.go:55]: This is a golang convention. I'd like to keep it.

File: ecdsa.go
Comment (shizhMSFT) [reply ecdsa.go:155]: I'd like to keep it for readability.

File: ed25519.go
Comment (shizhMSFT) [reply ed25519.go:44]: I'd like to keep it for readability.

File: sign.go
Comment (shizhMSFT) [reply sign.go:32]: @qmuntal Could you create an issue for this so that we can address later to unblock this PR.

File: cbor.go
Comment (shizhMSFT) [reply cbor.go:9]: Yes, they are intended for user consumption.

File: headers.go
Comment (shizhMSFT) [reply headers.go:47]: That's doable.

File: DESIGN.md
Comment (qmuntal) [reply DESIGN.md:183]: Sure!

**Review by qmuntal — APPROVED** [review]
Outstanding work! Approving with comments. I'll create issues for the comments that are not finally resolved so we can unblock this PR.

**Review by yogeshbdeshpande — APPROVED** [review]
Approve with a Caveat that the outstanding issues will be dealt with issue tracking in github followed by a new PR.

File: ecdsa.go
Comment (shizhMSFT) [reply ecdsa.go:97]: The following code shows that two implementations are not equivalent.

```go
package main

import (
	easn1 "encoding/asn1"
	"errors"
	"fmt"
	"math/big"

	"golang.org/x/crypto/cryptobyte"
	"golang.org/x/crypto/cryptobyte/asn1"
)

func main() {
	sig := encode()

	fmt.Println(decodeUsingCryptoByte(sig))
	fmt.Println(decodeUsingASN1Unmarshal(sig))
}

func encode() []byte {
	r := new(big.Int).SetInt64(42)
	s := new(big.Int).SetInt64(0)

	var b cryptobyte.Builder
	b.AddASN1(asn1.SEQUENCE, func(b *cryptobyte.Builder) {
		b.AddASN1BigInt(r)
		b.AddASN1BigInt(s)
		b.AddASN1OctetString([]byte("malicious")) // add some malicious content here
	})
	res, err := b.Bytes()
	if err != nil {
		panic(err)
	}
	return res
}

func decodeUsingCryptoByte(sig []byte) (r, s *big.Int, err error) {
	r, s = &big.Int{}, &big.Int{}
	var inner cryptobyte.String
	input := cryptobyte.String(sig)
	if !input.ReadASN1(&inner, asn1.SEQUENCE) ||
		!input.Empty() ||
		!inner.ReadASN1Integer(r) ||
		!inner.ReadASN1Integer(s) ||
		!inner.Empty() {
		return nil, nil, errors.New("ecdsa: invalid signature: invalid ASN.1 encoding")
	}
	return
}

func decodeUsingASN1Unmarshal(sig []byte) (r, s *big.Int, err error) {
	var esig struct {
		R, S *big.Int
	}
	if _, err := easn1.Unmarshal(sig, &esig); err != nil {
		return nil, nil, errors.New("ecdsa: invalid signature: invalid ASN.1 encoding")
	}
	return esig.R, esig.S, nil
}
```

The above code outputs:

```
<nil> <nil> ecdsa: invalid signature: invalid ASN.1 encoding
42 0 <nil>
```

Comment (shizhMSFT) [reply ecdsa.go:97]: However, since we are decoding from a trusted source just like the boringcrypto you have mentioned, I will re-implement this using the `asn1.Unmarhsal` method.

File: ecdsa.go
Comment (shizhMSFT) [reply ecdsa.go:23]: Good catch.

**Comment by shizhMSFT** [comment]
Issue #23 will be addressed in another PR.

**Review by SteveLasker — APPROVED** [review]
LGTM
Thanks to @shizhMSFT, @qmuntal, @thomas-fossati, @yogeshbdeshpande and the flurry of others that have been helping.

**Comment by SteveLasker** [comment]
Merging into [remote-signing](https://github.com/veraison/go-cose/tree/remote-signing) for additional stabilization and testing from Notary and others that would like to contribute and validate.

**Pull request merged**
